crates/ruff_linter/src/rules/ruff/rules/unnecessary_regular_expression.rs

```diff
@@ -173,9 +175,8 @@ impl<'a> ReFunc<'a> {
             // version
             ("sub", 3) => {
                 let repl = call.arguments.find_argument("repl", 1)?;
-                if !repl.is_string_literal_expr() {
-                    return None;
-                }
+                // make sure repl can be resolved to a string literal
+                resolve_string_literal(repl, semantic)?;
```

Comment:
If I understand correctly, I think the need here is slightly different to the need on lines 95-96. On lines 95-96, we do need to know the _value_ of the string in order to be able to check it doesn't have any metacharacters in it (so only a string literal will do, or something that we can resolve to a string literal). But here, we just need to know it's a string; any string will do, as long as the user _isn't_ passing in a function.

Is that the case? If so, it _might_ be worth adding back the `is_str` function you added in https://github.com/astral-sh/ruff/pull/14679/commits/efcc4cfbe121203443d146961f5fce2c9d117a69 and using that here, rather than using `resolve_string_literal` in both places. The advantage of the `is_str` technique is that it also understands basic type hints, e.g. it would understand that `re.sub()` is being passed a string for the `repl` argument in something like this:

```py
import re

def foo(input_str: str, repl: str):
    re.sub("foobar", repl, input_str)
```

Reply:
Now that I look at it, maybe we _do_ need to know the value of the `repl` string for a fully accurate analysis, though. For example, it seems like the initial version of the check that we merged yesterday emits a false-positive diagnostic for this:

```py
import re

re.sub(r"a", r"\g<0>\g<0>\g<0>", "a")
```

Now, this is a massive edge case -- I had to work quite hard to find it! It's _almost_ impossible to think of a way you can use a valid metacharacter in the replacement string without having any metacharacters in the pattern string. So maybe we shouldn't worry about this -- interested in your thoughts and @MichaReiser's!

Reply:
Oh wow, good catch! I was working on adding `is_str` back in, but maybe instead I need to check for metacharacters in `repl` too.

I thought we were safe from backreferences by avoiding `(` in the pattern, but I overlooked `\g<0>`. That exact sequence seems like the only way to trigger this behavior?

Reply:
> I thought we were safe from backreferences by avoiding `(` in the pattern, but I overlooked `\g<0>`. That exact sequence seems like the only way to trigger this behavior?

I think so, yes! Although we also emit a diagnostic on _invalid_ `re.sub()` calls like this, and maybe we should just ignore them? It feels like it might be outside of this rule's purview to autofix invalid `re.sub()` calls to valid `str.replace()` calls:

```pycon
>>> import re
>>> re.sub(r"a", r"\1", "a")
Traceback (most recent call last):
  File "<python-input-12>", line 1, in <module>
    re.sub(r"a", r"\1", "a")
    ~~~~~~^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/__init__.py", line 208, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/__init__.py", line 377, in _compile_template
    return _sre.template(pattern, _parser.parse_template(repl, pattern))
                                  ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/_parser.py", line 1070, in parse_template
    addgroup(int(this[1:]), len(this) - 1)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/_parser.py", line 1015, in addgroup
    raise s.error("invalid group reference %d" % index, pos)
re.PatternError: invalid group reference 1 at position 1
```

Reply:
I've added back `is_str` locally, along with your function argument test case. It's really nice to handle that case, but I'm a bit bothered by this edge case too, so I could go either way. I'm interested to hear which approach you and Micha think is best overall.

Reply:
Why don't you push the version with `is_str` to this PR, and we can see if it results in any more ecosystem hits? That might give us some more data on how useful it is to be able to detect that the `repl` argument is a string from the function annotation

Reply:
Hmm, it doesn't look like it adds any new ecosystem hits :/

I guess in that case, I'd vote for removing `is_str` again, and fixing the false positives on `\g<0>` and `\1` in `repl` arguments.

Thanks for putting up with my pernickitiness here!

Reply:
No problem, thanks for the thorough review! Should I reuse the other code to reject *any* metacharacters, or are references to named or numbered capture groups the only problems? I'm picturing checking for `\` followed by `g` or `1` through `9`. That seems a bit nicer than rejecting any metacharacter like I did for the patterns but possibly less safe.

Reply:
> I'm picturing checking for `\` followed by `g` or `1` through `9`. That seems a bit nicer than rejecting any metacharacter like I did for the patterns but possibly less safe.

I think actually we could check for `\` followed by any ASCII character _except_ one of `abfnrtv`. Other than `0-9` and `g` (which both have special behaviour in `repl` strings, as we've just been discussing!), I believe those are the _only_ ASCII escapes that will be permitted in a `repl` string by `re.sub()`, Anything else causes `re.PatternError` to be raised -- meaning it's probably out of scope for us to emit this diagnostic on it:

```pycon
>>> re.sub(r"a", r"\d", "a")
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    re.sub(r"a", r"\d", "a")
    ~~~~~~^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/__init__.py", line 208, in sub
    return _compile(pattern, flags).sub(repl, string, count)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/__init__.py", line 377, in _compile_template
    return _sre.template(pattern, _parser.parse_template(repl, pattern))
                                  ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.0/lib/python3.13/re/_parser.py", line 1076, in parse_template
    raise s.error('bad escape %s' % this, len(this)) from None
re.PatternError: bad escape \d at position 0
```

Reply:
Let me know what you think about this version. If it looks good, it might be nice to reuse this escape check for `pattern` as well instead of rejecting `\` entirely.